src/map/unit.cpp

```diff
@@ -78,6 +78,115 @@ struct unit_data* unit_bl2ud(struct block_list *bl)
+ * @return Whether the chase path was updated (true) or current movement can continue (false)
+ */
+bool unit_update_chase(block_list& bl, t_tick tick, bool fullcheck) {
+
+	unit_data* ud = unit_bl2ud(&bl);
+
+	if (ud == nullptr)
+		return true;
+
+	struct block_list* tbl = nullptr;
+	if (ud->target_to)
+		tbl = map_id2bl(ud->target_to);
+
+	// Reached destination, start attacking
+	if (tbl && tbl->m == bl.m && check_distance_bl(&bl, tbl, ud->chaserange) && status_check_visibility(&bl, tbl, false)) {
+		ud->to_x = bl.x;
+		ud->to_y = bl.y;
+		ud->target_to = 0;
+		// Aegis uses one before every attack, we should
+		// only need this one for syncing purposes.
+		clif_fixpos(bl);
+		if (ud->state.attack_continue)
+			unit_attack(&bl, tbl->id, ud->state.attack_continue);
+		return true;
+	}
+	// Cancel chase
+	else if (tbl == nullptr || (fullcheck && !status_check_visibility(&bl, tbl, (bl.type == BL_MOB)))) {
+		ud->to_x = bl.x;
+		ud->to_y = bl.y;
+
+		if (tbl != nullptr && bl.type == BL_MOB) {
+			mob_data& md = reinterpret_cast<mob_data&>(bl);
+			if (mob_warpchase(&md, tbl))
+				return true;
+			// Make sure monsters properly unlock their target, but still continue movement
+			mob_unlocktarget(&md, tick);
+			return false;
+		}
+
+		ud->target_to = 0;
+		return true;
+	}
+	// Update chase path
+	else if (fullcheck && ud->walkpath.path_pos > 0) {
+		unit_walktobl(&bl, tbl, ud->chaserange, ud->state.walk_easy | (ud->state.attack_continue ? 2 : 0));
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * Handles everything that happens when movement to the next cell is initiated
+ * @param bl: Moving bl
+ * @param sendMove: Whether move packet should be sent or not
+ * @param tick: Current tick
+ * @return Whether movement was initialized (true) or not (false)
+ */
+bool unit_walktoxy_nextcell(block_list& bl, bool sendMove, t_tick tick) {
+
+	unit_data* ud = unit_bl2ud(&bl);
+
+	if (ud == nullptr)
+		return true;
+
+	int32 speed;
+
+	if (ud->walkpath.path_pos >= ud->walkpath.path_len)
+		speed = -1;
+	else if (direction_diagonal(ud->walkpath.path[ud->walkpath.path_pos]))
+		speed = status_get_speed(&bl) * MOVE_DIAGONAL_COST / MOVE_COST;
+	else
+		speed = status_get_speed(&bl);
+
+	if (speed > 0) {
+		// Monsters check if their target is in range each cell
+		if (bl.type == BL_MOB && ud->target_to) {
+			short tx = ud->to_x;
+			short ty = ud->to_y;
+			// Monsters update their chase path one cell before reaching their final destination
+			if (unit_update_chase(bl, tick, (ud->walkpath.path_pos == ud->walkpath.path_len-1)))
+				return true;
+			// Continue moving, restore to_x and to_y
+			ud->to_x = tx;
+			ud->to_y = ty;
+		}
+
+		// Make sure there is no active walktimer
+		if (ud->walktimer != INVALID_TIMER) {
+			delete_timer(ud->walktimer, unit_walktoxy_timer);
+			ud->walktimer = INVALID_TIMER;
+		}
+		ud->walktimer = add_timer(tick + speed, unit_walktoxy_timer, bl.id, speed);
+		if (sendMove)
+			clif_move(*ud);
+		return true;
+	}
```

Comment:
Looks worse than it is.
Just removed 1 tab (see change above that makes the negative speed check obsolete) and changed short to int16
```suggestion
	// Monsters check if their target is in range each cell
	if (bl.type == BL_MOB && ud->target_to) {
		int16 tx = ud->to_x;
		int16 ty = ud->to_y;
		// Monsters update their chase path one cell before reaching their final destination
		if (unit_update_chase(bl, tick, (ud->walkpath.path_pos == ud->walkpath.path_len-1)))
			return true;
		// Continue moving, restore to_x and to_y
		ud->to_x = tx;
		ud->to_y = ty;
	}

	// Make sure there is no active walktimer
	if (ud->walktimer != INVALID_TIMER) {
		delete_timer(ud->walktimer, unit_walktoxy_timer);
		ud->walktimer = INVALID_TIMER;
	}
	ud->walktimer = add_timer(tick + speed, unit_walktoxy_timer, bl.id, speed);
	if (sendMove)
		clif_move(*ud);
	return true;
```

Reply:
Will apply this part myself (without setting speed to -1).

Reply:
Fixed, please see https://github.com/rathena/rathena/pull/8936/commits/abd715baa71756bb29ff30940816227bfa4c5bd6.